doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,258 @@
+
+```js
+/**
+ * @typedef {{
+ *  format: string
+ * }} ModuleExportsContext
+ */
+/**
+ * @typedef {{
+*   exports: any
+* }} ModuleExportsResult
+*/
+/**
+ * @param {string} url
+ * @param {ModuleExportsContext} context
+ * @param {(context: ModuleExportsContext) => {ModuleExportsResult}} nextExports
+ * @returns {ModuleExportsResult}
+ */
+function exports(url, context, nextExports) {  // Mocking require-in-the-middle
+  let { exports: originalExports } = nextExports(url, context);
+  const stats = getStats(url);
+  if (stats.name && modules.includes(stats.name)) {
+    const newExports = userCallback(originalExports, stats.name, stats.basedir);
+    return { exports: newExports }
+  }
+  return {
+    exports: originalExports
+  };
+}
+```
+
+This can only be meaningfully implemented for modules loaded by `require()`. ESM's design is completely different from CJS in that resolution of the graph and evaluation of the graph are separated, so a similar timing in ESM would be "after module evaluation". However, Node.js only gets to control the timing of the evaluation of the root module. The inner module evaluation is currently completely internal to V8. It may be possible to upstream a post-evaluation hook to V8, but calling from C++ to JS would also incur a non-trivial performance cost, especially if it needs to be done for every single module. Also, since the ESM namespace is specified to be immutable, what users can do after ESM evaluation is very limited - they cannot replace anything in the namespace or switch it to a different module. That's why the `link` hook was devised below to better work with the design of ESM (before linking completes it's possible to swap the module resolved by `import` to a different module).
+
+### Alternative design: `requires` that encompasses `resolve` and `load`
+
+An alternative design would be to span it across the `resolve` and `load` steps - take the specifier as argument, return exports in the result, and rename it to something like `requires()` (to avoid clashing with `require()`). The `nextRequires` hook would for example, invoke the default implementation of `requires` which in turn encompasses `resolve` and `load`. If `nextRequires` is not invoked then the default `resolve` and `load` will be skipped.
```

Comment:
We could have done this for the existing hooks too; that API originally had _four_ hooks, that we reduced to two when we added chaining. The idea behind resolve and load is that there are classes of customizations that you want to do that are only based on either resolving or only loading, like transpilation is generally only needed for `load`, and by splitting the two steps into separate hooks we allow for more possibilities of chaining where a resolve-only hook can run before any loading happens, etc.

Reply:
I think the higher level hooks are more useful when what the hooks want to customize are specifier to exports mappings (and the previous hooks didn’t provide post-evaluation customizations at all so that wouldn’t be obvious).

Reply:
(Or after the hooks were moved off-thread, the high-level hooks are no longer meaningful since modules are not transferrable, or even if they did, once they export any function, they are not transferrable to the main thread, so you can't really do any meaningful customizations post-load off-thread)